qiskit_experiments/data_processing/data_action.py

```diff
@@ -0,0 +1,76 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Defines the steps that can be used to analyse data."""
+
+from abc import ABCMeta, abstractmethod
+from typing import Any
+
+
+class DataAction(metaclass=ABCMeta):
+    """
+    Abstract action done on measured data to process it. Each subclass of DataAction must
+    define the way it formats, validates and processes data.
+    """
+
+    def __init__(self, validate: bool = True):
+        """
+        Args:
+            validate: If set to False the DataAction will not validate its input.
+        """
+        self._validate = validate
+
+    @abstractmethod
+    def _process(self, datum: Any) -> Any:
+        """
+        Applies the data processing step to the data.
+
+        Args:
+            datum: A single item of data which will be processed.
+
+        Returns:
+            processed data: The data that has been processed.
+        """
+
+    @abstractmethod
+    def _format_data(self, datum: Any, validate: bool = True) -> Any:
+        """
+        Check that the given data has the correct structure. This method may
+        additionally change the data type, e.g. converting a list to a numpy array.
+
+        Args:
+            datum: The data instance to check and format.
+            validate: If True the DataAction checks that the format of the datum is valid.
+
+        Returns:
+            datum: The data that was checked.
+
+        Raises:
+            DataProcessorError: If the data does not have the proper format.
+        """
+
+    def __call__(self, data: Any) -> Any:
```

Comment:
do we need call for `DataAction`? I think making `_process` public method is more readable, or `DataProcessor` can take arbitrary callback function (though they don't have validation method)?

Reply:
I think having callable nodes allows for a greater flexibility.